src/Media/ImageStorage.php

```diff
@@ -28,19 +28,22 @@ public function __construct(
 
     public function store(string $source, string $destination): void
     {
-        // Move to the local file system media directory
-        $this->localFilesystem->copy(
-            $source,
-            $this->mediaDirectory . '/' . $destination
-        );
-
-        // Upload to the S3 bucket
-        $this->s3FileSystem->writeStream(
-            $destination,
-            $this->localFilesystem->readStream($source),
-            [Config::OPTION_VISIBILITY => Visibility::PUBLIC]
-        );
-
+        try {
+            // Upload to the S3 bucket
+            $this->s3FileSystem->writeStream(
+                $destination,
+                $this->localFilesystem->readStream($source),
+                [Config::OPTION_VISIBILITY => Visibility::PUBLIC]
+            );
+        } catch (\Throwable $throwable) {
+            // Move to the local file system in case upload fails
+            $this->localFilesystem->copy(
+                $source,
+                $this->mediaDirectory . '/' . $destination
+            );
+            $this->localFilesystem->delete($source);
```

Comment:
The delete could go inside a `finally` block which will be executed always.

Reply:
I think yes https://stackoverflow.com/questions/41616790/how-is-the-keyword-finally-meant-to-be-used-in-php/41616978

Reply:
I tested it online http://sandbox.onlinephpfunctions.com/code/876bdbe8d689459af8d7a28d3bb116a307393acf